File: app/src/main/java/com/netspeed/monitor/SpeedIconGenerator.java
package com.netspeed.monitor;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.Typeface;

/**
 * Generates bitmap icons for the status bar notification.
 * Renders compact speed text onto a reusable bitmap for dynamic notification icons.
 * Uses static Paint/Bitmap objects to minimize allocations at high update rates.
 */
public final class SpeedIconGenerator {

    // Private constructor prevents instantiation
    private SpeedIconGenerator() {}

    // Icon canvas size in pixels — Android downscales for status bar
    private static final int SIZE = 130;

    // Font sizes for value line and unit line
    private static final float VALUE_TEXT_SIZE = 90f;
    private static final float UNIT_TEXT_SIZE = 52f;

    // Gap between value and unit text lines
    private static final float LINE_GAP = 2f;

    // Tighter letter spacing for compact rendering
    private static final float LETTER_SPACING = -0.05f;

    // Reusable bitmap — cleared and redrawn each call to avoid GC pressure
    private static Bitmap reusableBitmap;
    // Canvas attached to the reusable bitmap
    private static Canvas reusableCanvas;

    // Pre-allocated Paint for the large speed value number
    private static final Paint VALUE_PAINT = new Paint(Paint.ANTI_ALIAS_FLAG);
    // Pre-allocated Paint for the smaller unit label
    private static final Paint UNIT_PAINT = new Paint(Paint.ANTI_ALIAS_FLAG);
    // Reusable Rect objects for text measurement to avoid per-call allocation
    private static final Rect VALUE_BOUNDS = new Rect();
    private static final Rect UNIT_BOUNDS = new Rect();

    // Static initializer configures paints once at class load time
    static {
        Typeface typeface = Typeface.create("sans-serif-condensed", Typeface.BOLD);

        // Configure value paint — large bold number
        VALUE_PAINT.setColor(Color.WHITE);
        VALUE_PAINT.setTypeface(typeface);
        VALUE_PAINT.setTextAlign(Paint.Align.CENTER);
        VALUE_PAINT.setTextSize(VALUE_TEXT_SIZE);
        VALUE_PAINT.setLetterSpacing(LETTER_SPACING);

        // Configure unit paint — smaller unit label
        UNIT_PAINT.setColor(Color.WHITE);
        UNIT_PAINT.setTypeface(typeface);
        UNIT_PAINT.setTextAlign(Paint.Align.CENTER);
        UNIT_PAINT.setTextSize(UNIT_TEXT_SIZE);
        UNIT_PAINT.setLetterSpacing(LETTER_SPACING);
    }

    /**
     * Creates a bitmap icon showing the download speed.
     * Reuses a single bitmap and pre-allocated paints to stay efficient at 100ms update rates.
     */
    public static Bitmap createSpeedIcon(double downloadSpeed) {
        // Lazily create the reusable bitmap and canvas on first call
        if (reusableBitmap == null) {
            reusableBitmap = Bitmap.createBitmap(SIZE, SIZE, Bitmap.Config.ALPHA_8);
            reusableCanvas = new Canvas(reusableBitmap);
        }

        // Clear previous drawing — erase to transparent
        reusableBitmap.eraseColor(Color.TRANSPARENT);

        // Format speed into compact value + unit pair
        String[] formatted = SpeedUtils.formatSpeedCompact(downloadSpeed);
        String value = formatted[0];
        String unit = formatted[1];

        // Center X coordinate on the canvas
        float cx = SIZE / 2f;

        // Measure text bounds to center both lines vertically
        VALUE_PAINT.getTextBounds(value, 0, value.length(), VALUE_BOUNDS);
        UNIT_PAINT.getTextBounds(unit, 0, unit.length(), UNIT_BOUNDS);

        // Calculate vertical positions to center text block on canvas
        float totalHeight = VALUE_BOUNDS.height() + LINE_GAP + UNIT_BOUNDS.height();
        float startY = (SIZE - totalHeight) / 2f;

        // Draw value text on the top line
        float valueY = startY + VALUE_BOUNDS.height();
        reusableCanvas.drawText(value, cx, valueY, VALUE_PAINT);

        // Draw unit text on the bottom line
        float unitY = valueY + LINE_GAP + UNIT_BOUNDS.height();
        reusableCanvas.drawText(unit, cx, unitY, UNIT_PAINT);

        return reusableBitmap;
    }
}

File: app/src/main/java/com/netspeed/monitor/service/SpeedIconGenerator.kt
package com.netspeed.monitor.service

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import android.graphics.Typeface
import androidx.core.graphics.createBitmap

object SpeedIconGenerator {

    // ====================================================
    // ALL CONSTANTS - ADJUST FROM HERE
    // ====================================================

    // Icon canvas size in pixels — larger canvas = more detail before Android downscales
    private const val SIZE = 130

    // Font size for speed value (top line: "1.2", "23", etc.)
    private const val VALUE_TEXT_SIZE = 90f

    // Font size for unit label (bottom line: "KB/s", "MB/s", etc.)
    private const val UNIT_TEXT_SIZE = 52f

    // Gap between value and unit lines (pixels)
    private const val LINE_GAP = 2f

    // Letter spacing (negative = tighter)
    private const val LETTER_SPACING = -0.05f

    // Text color
    private const val TEXT_COLOR = Color.WHITE

    // Font family
    private const val FONT_FAMILY = "sans-serif-condensed"

    // Font style: Typeface.BOLD, Typeface.NORMAL, Typeface.ITALIC, Typeface.BOLD_ITALIC
    private val FONT_STYLE = Typeface.BOLD

    // ====================================================

    fun createDualSpeedIcon(
        downloadSpeed: Double,
        @Suppress("UNUSED_PARAMETER") uploadSpeed: Double
    ): Bitmap {
        val bitmap = createBitmap(SIZE, SIZE)
        val canvas = Canvas(bitmap)

        val (value, unit) = formatCompact(downloadSpeed)

        val typeface = Typeface.create(FONT_FAMILY, FONT_STYLE)
        val cx = SIZE / 2f

        val valuePaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            color = TEXT_COLOR
            this.typeface = typeface
            textAlign = Paint.Align.CENTER
            textSize = VALUE_TEXT_SIZE
            letterSpacing = LETTER_SPACING
        }

        val unitPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            color = TEXT_COLOR
            this.typeface = typeface
            textAlign = Paint.Align.CENTER
            textSize = UNIT_TEXT_SIZE
            letterSpacing = LETTER_SPACING
        }

        val valueBounds = Rect()
        valuePaint.getTextBounds(value, 0, value.length, valueBounds)

        val unitBounds = Rect()
        unitPaint.getTextBounds(unit, 0, unit.length, unitBounds)

        val totalHeight = valueBounds.height() + LINE_GAP + unitBounds.height()
        val startY = (SIZE - totalHeight) / 2f

        val valueY = startY + valueBounds.height()
        canvas.drawText(value, cx, valueY, valuePaint)

        val unitY = valueY + LINE_GAP + unitBounds.height()
        canvas.drawText(unit, cx, unitY, unitPaint)

        return bitmap
    }

    fun createSpeedIcon(downloadSpeed: Double): Bitmap =
        createDualSpeedIcon(downloadSpeed, 0.0)

    private fun formatCompact(bytesPerSecond: Double): Pair<String, String> {
        return when {
            // 1 GB+ -> "1", "2", etc.
            bytesPerSecond >= 1_073_741_824 ->
                Pair(String.format("%.0f", bytesPerSecond / 1_073_741_824), "GB/s")

            // 100 MB+ -> "100", "234", etc.
            bytesPerSecond >= 104_857_600 ->
                Pair(String.format("%.0f", bytesPerSecond / 1_048_576), "MB/s")

            // 10 MB+ -> "10", "56", etc.
            bytesPerSecond >= 10_485_760 ->
                Pair(String.format("%.0f", bytesPerSecond / 1_048_576), "MB/s")

            // 1 MB+ -> "1.2", "5.8", etc.
            bytesPerSecond >= 1_048_576 ->
                Pair(String.format("%.1f", bytesPerSecond / 1_048_576), "MB/s")

            // 100 KB+ -> "100", "456", etc.
            bytesPerSecond >= 102_400 ->
                Pair(String.format("%.0f", bytesPerSecond / 1024), "KB/s")

            // 1 KB+ -> "1.2", "45.6", etc.
            bytesPerSecond >= 1024 ->
                Pair(String.format("%.1f", bytesPerSecond / 1024), "KB/s")

            // Less than 1 KB -> "512", "64", etc.
            bytesPerSecond > 0 ->
                Pair(String.format("%.0f", bytesPerSecond), "B/s")

            // Zero
            else -> Pair("0", "B/s")
        }
    }
}